samples/led-matrix-weather/Program.cs

```diff
@@ -0,0 +1,274 @@
+        {
+            if (args.Length != 1)
+            {
+                Console.WriteLine("Usage: led-matrix-weather <openweathermap.org API key>");
+                Console.WriteLine();
+                Console.WriteLine("If wrong key is used or there is no network connection");
+                Console.WriteLine("example data will be displayed.");
+                Console.WriteLine();
+                Console.WriteLine("For the first 30 seconds of execution");
+                Console.WriteLine("IP addresses will be displayed instead of data.");
+                return;
+            }
+
+            s_client = new Weather(args[0]);
+
+            UpdateWeather();
+            Task.Run(WeatherUpdater);
+            Task.Run(IpsGetter);
+
+            PinMapping mapping = PinMapping.MatrixBonnetMapping32;
+            RGBLedMatrix matrix = new RGBLedMatrix(mapping, 64, 64, 2, 2);
+
+            Task drawing = Task.Run(() =>
+            {
+                matrix.StartRendering();
+
+                while (true)
+                {
+                    Action<RGBLedMatrix> scenario = s_scenario;
+
+                    if (scenario == null)
+                        break;
+
+                    Stopwatch sw = Stopwatch.StartNew();
+                    scenario(matrix);
+
+                    if (s_scenario != null && sw.ElapsedMilliseconds < 100)
+                    {
+                        Debug.WriteLine("Scenario execution finished in less than 100ms. This is likely due to bug.");
+                    }
+                }
+            });
+
+            try
+            {
+                if (!Console.IsOutputRedirected)
+                {
+                    while (s_scenario != null)
+                    {
+                        switch (Console.ReadKey(intercept: true).Key)
+                        {
+                            case ConsoleKey.Q:
+                            {
+                                s_scenario = null;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                drawing.Wait();
+            }
+            finally
+            {
+                matrix.Dispose();
+            }
+        }
+
+        static void WeatherUpdater()
+        {
+            while (true)
+            {
+                UpdateWeather();
+                Thread.Sleep(5000);
+            }
+        }
+
+        static void IpsGetter()
+        {
+            try
+            {
+                s_ips = GetLocalNetworkIPAddresses().ToArray();
+            }
+            catch
+            {
+                // Choke all errors.
+                // This is function is used for diagnostics
+                // we don't want it to interrupt the process in any way
+                s_ips = new string[1] { "Cannot get IPs" };
+            }
+
+            s_showLocalIp = Stopwatch.StartNew();
+        }
+
+        static void UpdateWeather()

```

Comment:
Why make this a void instead of a OpenWeatherResponse?

Reply:
it's hard to name when this returns the value. I.e.:
`bool GetLatestWeatherInfoOrExampleData(out OpenWeatherResponse)` where true would mean that using real data and false example but then bool argument is meaningless unless you read the code so needs docs etc. Current version is just simpler to read since there is only one caller anyway (actually two but used in the same context)